# < 간단한 몇 가지 >

## → array, list, dictionary로 바꾸기

In [1]:
import pydataset
mpg = pydataset.data('mpg')

df = mpg.iloc[:5, 2:5]
df

displ  year  cyl
1    1.8  1999    4
2    1.8  1999    4
3    2.0  2008    4
4    2.0  2008    4
5    2.8  1999    6

In [2]:
#1. df를 array로 바꾸기
df.values

array([[1.800e+00, 1.999e+03, 4.000e+00],
       [1.800e+00, 1.999e+03, 4.000e+00],
       [2.000e+00, 2.008e+03, 4.000e+00],
       [2.000e+00, 2.008e+03, 4.000e+00],
       [2.800e+00, 1.999e+03, 6.000e+00]])

In [3]:
#2. df를 딕셔너리로 바꾸기
df.to_dict()

{'displ': {1: 1.8, 2: 1.8, 3: 2.0, 4: 2.0, 5: 2.8},
 'year': {1: 1999, 2: 1999, 3: 2008, 4: 2008, 5: 1999},
 'cyl': {1: 4, 2: 4, 3: 4, 4: 4, 5: 6}}

In [4]:
#3. df를 리스트로 바꾸기
df.values.tolist() #리스트로 바꾸려면 데이터프레임일때는 X

[[1.8, 1999.0, 4.0],
 [1.8, 1999.0, 4.0],
 [2.0, 2008.0, 4.0],
 [2.0, 2008.0, 4.0],
 [2.8, 1999.0, 6.0]]

## → 리스트를 가로 방향으로 한 줄 세우기

In [5]:
#df를 하나의 리스트로 만들기(가로로 줄 세우기)
li = []
for i in df.values.tolist():    #정석 방식
    for j in i:
        li.append(j)
li

[1.8,
 1999.0,
 4.0,
 1.8,
 1999.0,
 4.0,
 2.0,
 2008.0,
 4.0,
 2.0,
 2008.0,
 4.0,
 2.8,
 1999.0,
 6.0]

In [6]:
df.values.reshape(-1).tolist()   #같은방법

[1.8,
 1999.0,
 4.0,
 1.8,
 1999.0,
 4.0,
 2.0,
 2008.0,
 4.0,
 2.0,
 2008.0,
 4.0,
 2.8,
 1999.0,
 6.0]

In [7]:
df.values.flatten().tolist() #flatten() '평평하게 편다'

[1.8,
 1999.0,
 4.0,
 1.8,
 1999.0,
 4.0,
 2.0,
 2008.0,
 4.0,
 2.0,
 2008.0,
 4.0,
 2.8,
 1999.0,
 6.0]

## → map 이용하여 딕셔너리 이름 할당하기

In [6]:
from sklearn.datasets import load_iris
iris = load_iris()

iris_data = iris['data']
iris_label = iris['target']

#데이터 프레임으로 변환하기
iris_df = pd.DataFrame(data = iris_data, columns=iris['feature_names'])

#새로운 칼럼 만들기.(label)
iris_df['label'] = iris['target']

iris_df

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)  \
0                  5.1               3.5                1.4               0.2   
1                  4.9               3.0                1.4               0.2   
2                  4.7               3.2                1.3               0.2   
3                  4.6               3.1                1.5               0.2   
4                  5.0               3.6                1.4               0.2   
..                 ...               ...                ...               ...   
145                6.7               3.0                5.2               2.3   
146                6.3               2.5                5.0               1.9   
147                6.5               3.0                5.2               2.0   
148                6.2               3.4                5.4               2.3   
149                5.9               3.0                5.1               1.8   

     label  
0        0  
1        0  
2        0  
3        0  
4        0  
..     ...  
145      2  
146      2  
147      2  
148      2  
149      2  

[150 rows x 5 columns]

In [8]:
#map을 사용하여 딕셔너리로 이름 할당하기.
iris_df['label'] = iris_df['label'].map({0:'setosa', 1:'versicolior', 2:'virginica'})
iris_df
# iris_df['label'].map(label_mapping)도 같은 방식

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)  \
0                  5.1               3.5                1.4               0.2   
1                  4.9               3.0                1.4               0.2   
2                  4.7               3.2                1.3               0.2   
3                  4.6               3.1                1.5               0.2   
4                  5.0               3.6                1.4               0.2   
..                 ...               ...                ...               ...   
145                6.7               3.0                5.2               2.3   
146                6.3               2.5                5.0               1.9   
147                6.5               3.0                5.2               2.0   
148                6.2               3.4                5.4               2.3   
149                5.9               3.0                5.1               1.8   

    label  
0     NaN  
1     NaN  
2     NaN  
3     NaN  
4     NaN  
..    ...  
145   NaN  
146   NaN  
147   NaN  
148   NaN  
149   NaN  

[150 rows x 5 columns]

# < 사이킷런 >

In [8]:
import sklearn

## □ 붓꽃 품종 예측하기 

### 1. 데이터 세트 분리 : train_test_split( )
: 데이터를 학습 데이터와 테스트 데이터로 분리합니다.

In [35]:
from sklearn.datasets import load_iris
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import train_test_split  #분리

iris = load_iris()
iris

{'data': array([[5.1, 3.5, 1.4, 0.2],
        [4.9, 3. , 1.4, 0.2],
        [4.7, 3.2, 1.3, 0.2],
        [4.6, 3.1, 1.5, 0.2],
        [5. , 3.6, 1.4, 0.2],
        [5.4, 3.9, 1.7, 0.4],
        [4.6, 3.4, 1.4, 0.3],
        [5. , 3.4, 1.5, 0.2],
        [4.4, 2.9, 1.4, 0.2],
        [4.9, 3.1, 1.5, 0.1],
        [5.4, 3.7, 1.5, 0.2],
        [4.8, 3.4, 1.6, 0.2],
        [4.8, 3. , 1.4, 0.1],
        [4.3, 3. , 1.1, 0.1],
        [5.8, 4. , 1.2, 0.2],
        [5.7, 4.4, 1.5, 0.4],
        [5.4, 3.9, 1.3, 0.4],
        [5.1, 3.5, 1.4, 0.3],
        [5.7, 3.8, 1.7, 0.3],
        [5.1, 3.8, 1.5, 0.3],
        [5.4, 3.4, 1.7, 0.2],
        [5.1, 3.7, 1.5, 0.4],
        [4.6, 3.6, 1. , 0.2],
        [5.1, 3.3, 1.7, 0.5],
        [4.8, 3.4, 1.9, 0.2],
        [5. , 3. , 1.6, 0.2],
        [5. , 3.4, 1.6, 0.4],
        [5.2, 3.5, 1.5, 0.2],
        [5.2, 3.4, 1.4, 0.2],
        [4.7, 3.2, 1.6, 0.2],
        [4.8, 3.1, 1.6, 0.2],
        [5.4, 3.4, 1.5, 0.4],
        [5.2, 4.1, 1.5, 0.1],
  

In [13]:
type(iris) 
#Bunch라고 출력되지만 사실은 딕셔너리와 같은것

sklearn.utils.Bunch

In [ ]:
iris.keys()
iris.values()
iris.items() #key와 values를 동시에 보여준다.

In [15]:
iris_data = iris['data']
iris_data  #딕셔너리에서 특정 값을 찾는 것과 같은 방식

array([[5.1, 3.5, 1.4, 0.2],
       [4.9, 3. , 1.4, 0.2],
       [4.7, 3.2, 1.3, 0.2],
       [4.6, 3.1, 1.5, 0.2],
       [5. , 3.6, 1.4, 0.2],
       [5.4, 3.9, 1.7, 0.4],
       [4.6, 3.4, 1.4, 0.3],
       [5. , 3.4, 1.5, 0.2],
       [4.4, 2.9, 1.4, 0.2],
       [4.9, 3.1, 1.5, 0.1],
       [5.4, 3.7, 1.5, 0.2],
       [4.8, 3.4, 1.6, 0.2],
       [4.8, 3. , 1.4, 0.1],
       [4.3, 3. , 1.1, 0.1],
       [5.8, 4. , 1.2, 0.2],
       [5.7, 4.4, 1.5, 0.4],
       [5.4, 3.9, 1.3, 0.4],
       [5.1, 3.5, 1.4, 0.3],
       [5.7, 3.8, 1.7, 0.3],
       [5.1, 3.8, 1.5, 0.3],
       [5.4, 3.4, 1.7, 0.2],
       [5.1, 3.7, 1.5, 0.4],
       [4.6, 3.6, 1. , 0.2],
       [5.1, 3.3, 1.7, 0.5],
       [4.8, 3.4, 1.9, 0.2],
       [5. , 3. , 1.6, 0.2],
       [5. , 3.4, 1.6, 0.4],
       [5.2, 3.5, 1.5, 0.2],
       [5.2, 3.4, 1.4, 0.2],
       [4.7, 3.2, 1.6, 0.2],
       [4.8, 3.1, 1.6, 0.2],
       [5.4, 3.4, 1.5, 0.4],
       [5.2, 4.1, 1.5, 0.1],
       [5.5, 4.2, 1.4, 0.2],
       [4.9, 3

In [16]:
iris_label = iris['target']
iris_label

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2])

In [1]:
X_train, X_test, y_train, y_test = train_test_split(iris_data, iris_label)
# _train은 학습용 데이터, _test은 테스트용 데이터를 나타낸다.
# iris_data -> X값   iris_label -> y값

NameError: name 'train_test_split' is not defined

In [21]:
X_train.shape

(112, 4)

데이터의 갯수는 112개라는 이야기.

칼럼.특질.피쳐는 4개이다.

#### → 주요 Key의 의미

```
- data : X(피처)의 데이터 세트

- target : 최종적으로 구하려는 결괏값
           (분류 시에는 레이블 값, 회귀 시에는 숫자 결괏값 데이터 세트로 나타난다.)

- target_names : 개별 y(레이블)의 이름, 분류하고자 하는 대상

- feature_names : 피처의 이름

- DESCR : describe의 약자로 데이터에 대한 설명
```

#### → train_test_split() Parameter

In [ ]:
from sklearn.model_selection import train_test_split
train_test_split(arrays, test_size, train_size, random_state, shuffle, stratify)

```
- arrays : 분할시킬 데이터를 입력

- test_size : 테스트 데이터셋의 비율(float)이나 갯수(int) (default = 0.25)

- train_size : 학습 데이터셋의 비율(float)이나 갯수(int) (default = test_size의 나머지)

- random_state : 데이터 분할시 셔플이 이루어지는데 이를 위한 시드값 (int나 RandomState로 입력)

- shuffle : 셔플여부설정 (default = True)

- stratify : 지정한 Data의 비율을 유지한다. 
             예를 들어, Label Set인 Y가 25%의 0과 75%의 1로 이루어진 Binary Set일 때, 
             stratify=Y로 설정하면 나누어진 데이터셋들도 0과 1을 각각 25%, 75%로 유지한 채                분할된다.
```

### 2. 모델 학습
: 학습 데이터를 기반으로 ML 알고리즘을 적용해 모델을 학습시킵니다.

#### 1) DecisionTreeClassifier( ) 
: 객체 생성

In [36]:
dt_clf = DecisionTreeClassifier()
# DecisionTreeClassifier 객체 생성(아직 아무런 기능 없다.)

#### 2) fit( )

In [20]:
dt_clf.fit(X_train, y_train)
#fitting 적합한다. 학습한다. 훈련한다.

DecisionTreeClassifier()

DecisionTreeClassifier 객체는 학습 데이터를 기반으로 학습이 완료됌.

이제 이 객체의 predict()에 테스트 데이터를 이용하여 예측값을 반환함.

In [32]:
X_test.shape  #38개가 있다.

(38, 4)

### 3. 예측 수행 : predict( )
: 학습된 ML 모델을 이용해 테스트 데이터의 분류(즉, 붓꽃 종류)를 예측합니다.

In [23]:
y_pred = dt_clf.predict(X_test)
y_pred  #예측

array([2, 1, 0, 2, 0, 0, 2, 2, 1, 1, 0, 2, 2, 1, 0, 2, 0, 1, 2, 0, 2, 1,
       1, 1, 2, 1, 1, 0, 1, 0, 2, 1, 0, 0, 2, 0, 0, 1])

In [24]:
y_test  #정답

array([1, 1, 0, 2, 0, 0, 2, 2, 1, 1, 0, 2, 2, 1, 0, 2, 0, 1, 2, 0, 2, 2,
       1, 1, 2, 1, 1, 0, 1, 0, 2, 1, 0, 0, 2, 0, 0, 1])

In [33]:
y_pred == y_test  # 둘 간 얼마나 일치한가?

array([False,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True, False,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True])

In [34]:
sum(y_pred == y_test) #일치 갯수는 36개, 처음에는 38개였다.

36

In [28]:
iris.target_names

array(['setosa', 'versicolor', 'virginica'], dtype='<U10')

위에 y_test 출력값을 보면 1,0 이런식으로 되어있는데

이 숫자가 의미하는 바는 위에 어레이의 인덱스를 표현한 것이다.

### 4. 평가 : accuracy_score( )
: 이렇게 예측된 결괏값과 테스트 데이터의 실제 결괏값을 비교해 ML 모델 성능을 평가한다.

In [30]:
from sklearn.metrics import accuracy_score  #정확도를 보여준다.

In [31]:
accuracy_score(y_test, y_pred)

0.9473684210526315

# < 데이터 전처리 >
```
< 이유 >

- 결손값, 즉 NaN, Null 값은 허용되지 않습니다. (이러한 경우 pandas로 처리)
- 문자열 값을 입력값으로 허용하지 않습니다. (그래서 0,1 이런식으로 표현함.)

```

## □ 레이블 인코딩: LabelEncoder
- LabelEncoder를 객체로 생성 후 fit()과 transform()을 호출해 인코딩

In [39]:
from sklearn.preprocessing import LabelEncoder

encoder = LabelEncoder()

In [40]:
labels = encoder.fit_transform(['TV', '냉장고', '전자레인지'])
labels

array([0, 1, 2], dtype=int64)

## □ 원-핫 인코딩
One-Hot Encoding / '한 개만 불이 들어와요'

### 1) 데이터 불러오기

In [9]:
from sklearn.preprocessing import OneHotEncoder
import numpy as np

items = ['TV','냉장고','전자레인지','컴퓨터','선풍기','선풍기','믹서','믹서']
items

['TV', '냉장고', '전자레인지', '컴퓨터', '선풍기', '선풍기', '믹서', '믹서']

### 2) 2차원 데이터로 변환

In [10]:
items = np.array(items).reshape(-1,1)
# 한줄로 쭉 서있는게 아니라 세로로 한줄로 들어가야한다.(열이 1개)
items

array([['TV'],
       ['냉장고'],
       ['전자레인지'],
       ['컴퓨터'],
       ['선풍기'],
       ['선풍기'],
       ['믹서'],
       ['믹서']], dtype='<U5')

### 3) OneHotEncoder 클래스로 변환

In [11]:
encoder = OneHotEncoder()
one_hot_labels = encoder.fit_transform(items)
one_hot_labels

<8x6 sparse matrix of type '<class 'numpy.float64'>'
	with 8 stored elements in Compressed Sparse Row format>

Sparse -> 희소행렬

### 4) toarray( )
:희소행렬 형태를 밀집행렬로 변환하기

In [12]:
one_hot_labels.toarray()

array([[1., 0., 0., 0., 0., 0.],
       [0., 1., 0., 0., 0., 0.],
       [0., 0., 0., 0., 1., 0.],
       [0., 0., 0., 0., 0., 1.],
       [0., 0., 0., 1., 0., 0.],
       [0., 0., 0., 1., 0., 0.],
       [0., 0., 1., 0., 0., 0.],
       [0., 0., 1., 0., 0., 0.]])

### ▶ 또 다른 방법 : get_dummies( )
: array로 바꿀 필요 없이 API 사용하면 간단하다.

In [13]:
items = ['TV','냉장고','전자레인지','컴퓨터','선풍기','선풍기','믹서','믹서']
pd.get_dummies(items)

TV  냉장고  믹서  선풍기  전자레인지  컴퓨터
0   1    0   0    0      0    0
1   0    1   0    0      0    0
2   0    0   0    0      1    0
3   0    0   0    0      0    1
4   0    0   0    1      0    0
5   0    0   0    1      0    0
6   0    0   1    0      0    0
7   0    0   1    0      0    0

# < 교차검증 >
- Cross Validation
- 크로스는 여러번 이라는 뜻
- 교차검증은 여러번 검증한다는 것(여러 경로는 통해서 검증)

## ▶ KFlod(n_splits = )

In [14]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score
from sklearn.model_selection import KFold
import numpy as np

iris = load_iris()                   # 파일부르기

iris_data = iris.data                # data(정보)와 target(매치대상) 저장
iris_label= iris.target

dt_clf=DecisionTreeClassifier()      # 의사결정 트리 

kfold = KFold(n_splits = 5)          # 5번 검증한다

In [15]:
cv_accuracy = []
n_iter = 0

for train_index, test_index in kfold.split(iris_data):
    print(train_index, test_index)
    print("---"*10)

[ 30  31  32  33  34  35  36  37  38  39  40  41  42  43  44  45  46  47
  48  49  50  51  52  53  54  55  56  57  58  59  60  61  62  63  64  65
  66  67  68  69  70  71  72  73  74  75  76  77  78  79  80  81  82  83
  84  85  86  87  88  89  90  91  92  93  94  95  96  97  98  99 100 101
 102 103 104 105 106 107 108 109 110 111 112 113 114 115 116 117 118 119
 120 121 122 123 124 125 126 127 128 129 130 131 132 133 134 135 136 137
 138 139 140 141 142 143 144 145 146 147 148 149] [ 0  1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18 19 20 21 22 23
 24 25 26 27 28 29]
------------------------------
[  0   1   2   3   4   5   6   7   8   9  10  11  12  13  14  15  16  17
  18  19  20  21  22  23  24  25  26  27  28  29  60  61  62  63  64  65
  66  67  68  69  70  71  72  73  74  75  76  77  78  79  80  81  82  83
  84  85  86  87  88  89  90  91  92  93  94  95  96  97  98  99 100 101
 102 103 104 105 106 107 108 109 110 111 112 113 114 115 116 117 118 119
 120 121 122 123 124 12

```
2개중 첫번째 인덱스는 4개에 대한 인덱스(train_data)
뒤에는 나머지 한개의 인덱스(test_data)
이렇게 1덩어리가 5개로 쪼개져있다.
----------------------------------------
0,1,2가 있는데 0,1은 50개씩 있는데 2가 5개밖에 없으면 train이나 test에 편중되어서
2가 없는 상태로 학습이 되어서 맞추지 못하는 상황이 생긴다.

즉, 비율을 맞춰야 한다.

0이 100개 1은 40개 2가 10로 편중된 데이터가 있다면 
100,40,10개 각각 7:3비율로 7은 train으로 3은 test로 골고루 분산되게끔 만들어야한다.

test_size = 0.3
```

In [20]:
iris_data = iris.data
                # data(정보)와 target(매치대상) 저장
iris_label= iris.target


dt_clf = DecisionTreeClassifier()      # 의사결정 트리 
kfold = KFold(n_splits=5)

cv_accuracy = []

# KFold객체의 split( ) 호출하면 폴드 별 학습용, 검증용 테스트의 인덱스를 반환한다. 이거 중요함.  
for train_index, test_index  in kfold.split(iris_data): 
    
    # print(train_index, test_index)
    # print("---"*10)
    # kfold.split( )으로 반환된 인덱스를 이용하여 학습용, 검증용 테스트 데이터 추출
    X_train, X_test = iris_data[train_index], iris_data[test_index]
    y_train, y_test = iris_label[train_index], iris_label[test_index]

    
    dt_clf.fit(X_train , y_train)    
    pred = dt_clf.predict(X_test)
    
   
    accuracy = accuracy_score(y_test,pred)   
    cv_accuracy.append(accuracy)
 
    print("개별 검증 정확도: ", cv_accuracy)
    
print('평균 검증 정확도:', np.mean(cv_accuracy))

개별 검증 정확도:  [1.0]
개별 검증 정확도:  [1.0, 1.0]
개별 검증 정확도:  [1.0, 1.0, 0.8333333333333334]
개별 검증 정확도:  [1.0, 1.0, 0.8333333333333334, 0.9333333333333333]
개별 검증 정확도:  [1.0, 1.0, 0.8333333333333334, 0.9333333333333333, 0.7666666666666667]
평균 검증 정확도: 0.9066666666666666


p102, 103 줄인 모습

```
<< 여기서 split가 중요한 이유 >>

- split는 입력값이 데이터프레임이 들어간다.
- 그럼 출력값은 똑같이 데이터프레임인가? 아니다. 데이터프레임 인덱스를 출력하게 된다.
- 해당 인덱스를 넣으면 거기에 해당하는 값이 출력하게 되는데 
  그 원리를 이용한게 저 부분이다.
```

### → StratifiedKFold : 제대로 분배X일때

In [17]:
from sklearn.datasets import load_iris #데이터 불러오기
from sklearn.tree import DecisionTreeClassifier  #정해주기
from sklearn.metrics import accuracy_score #평가기준
from sklearn.model_selection import StratifiedKFold #층화된 kfold를 가져온다.

iris = load_iris()
features = X = iris_data = iris.data
labels = y = iris_label = iris.target

dt_clf = DecisionTreeClassifier()

s_kfold = StratifiedKFold(n_splits = 3)
cv_accuracy = []

for train_index, test_index in s_kfold.split(features, labels): #비율 다 감안해서 쪼갠것
    print(train_index)
    print(test_index)
    print()

[ 17  18  19  20  21  22  23  24  25  26  27  28  29  30  31  32  33  34
  35  36  37  38  39  40  41  42  43  44  45  46  47  48  49  67  68  69
  70  71  72  73  74  75  76  77  78  79  80  81  82  83  84  85  86  87
  88  89  90  91  92  93  94  95  96  97  98  99 116 117 118 119 120 121
 122 123 124 125 126 127 128 129 130 131 132 133 134 135 136 137 138 139
 140 141 142 143 144 145 146 147 148 149]
[  0   1   2   3   4   5   6   7   8   9  10  11  12  13  14  15  16  50
  51  52  53  54  55  56  57  58  59  60  61  62  63  64  65  66 100 101
 102 103 104 105 106 107 108 109 110 111 112 113 114 115]

[  0   1   2   3   4   5   6   7   8   9  10  11  12  13  14  15  16  34
  35  36  37  38  39  40  41  42  43  44  45  46  47  48  49  50  51  52
  53  54  55  56  57  58  59  60  61  62  63  64  65  66  83  84  85  86
  87  88  89  90  91  92  93  94  95  96  97  98  99 100 101 102 103 104
 105 106 107 108 109 110 111 112 113 114 115 133 134 135 136 137 138 139
 140 141 142 143 144 14

In [18]:
from sklearn.datasets import load_iris #데이터 불러오기
from sklearn.tree import DecisionTreeClassifier  #정해주기
from sklearn.metrics import accuracy_score #평가기준
from sklearn.model_selection import StratifiedKFold #층화된 kfold를 가져온다.

iris = load_iris()
features = X = iris_data = iris.data
labels = y = iris_label = iris.target

dt_clf = DecisionTreeClassifier()

s_kfold = StratifiedKFold(n_splits = 3)
cv_accuracy = []

for train_index, test_index in s_kfold.split(features, labels):

    X_train, X_test = features[train_index], features[test_index]
    y_train, y_test = labels[train_index], labels[test_index]
    
    dt_clf.fit(X_train, y_train)
    pred = dt_clf.predict(X_test)
    
    acc = accuracy_score(y_test, pred)  #이 두 개를 비교해서 정확도 보가
    cv_accuracy.append(acc)
    print("개별 정확도: ", acc)
    
print('정확도 모음: ', cv_accuracy)

개별 정확도:  0.98
개별 정확도:  0.94
개별 정확도:  0.96
정확도 모음:  [0.98, 0.94, 0.96]


In [19]:
np.mean(cv_accuracy).round(4)

0.96

## ▶ cross_val_score( ) : for문 보다 간단

In [21]:
from sklearn.model_selection import cross_val_score, cross_validate #교차검증 점수

iris = load_iris()
data = features = X = iris_data = iris.data
label = labels = y = iris_label = iris.target

dt_clf = DecisionTreeClassifier()

In [22]:
scores = cross_val_score(dt_clf, data, label, scoring = 'accuracy', cv = 3) #3번 쪼개기
scores

array([0.98, 0.92, 0.96])

In [23]:
np.mean(scores).round(4)

0.9533

# < GridSearchCV : 최적 추출 >
: 최적 하이퍼 파라미터 추출하기

## 1) max_depth, min_samples_split

In [1]:
from sklearn.datasets import load_iris
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score
from sklearn.model_selection import GridSearchCV, train_test_split

In [5]:
iris_data = iris.data
iris_label= iris.target

iris = load_iris()

X_train, X_test, y_train, y_test = train_test_split(iris_data, iris_label, test_size = 0.2, 
                                                    random_state=121) 

dt_clf = DecisionTreeClassifier()

grid_parameters = {'max_depth':[1,2,3],
                  'min_samples_split': [2,3]}

grid_tree = GridSearchCV(dt_clf, param_grid=grid_parameters, cv=3, refit=True)

grid_tree.fit(X_train, y_train) #학습이 이루어짐

GridSearchCV(cv=3, estimator=DecisionTreeClassifier(),
             param_grid={'max_depth': [1, 2, 3], 'min_samples_split': [2, 3]})

### → cv_results_

```
cv_results_는 GridSearchcv의 결과 세트로서 딕셔너리 형태로 key값과 리스트 형태인 value값을 가진다.

cv_results_는 데이터프레임으로 변환해야 보기 편하다.
```

In [7]:
import pandas as pd

pd.DataFrame(grid_tree.cv_results_).iloc[:,5:] 
#보기 좋게 데이터프레임으로 변경함

param_min_samples_split                                    params  \
0                       2  {'max_depth': 1, 'min_samples_split': 2}   
1                       3  {'max_depth': 1, 'min_samples_split': 3}   
2                       2  {'max_depth': 2, 'min_samples_split': 2}   
3                       3  {'max_depth': 2, 'min_samples_split': 3}   
4                       2  {'max_depth': 3, 'min_samples_split': 2}   
5                       3  {'max_depth': 3, 'min_samples_split': 3}   

   split0_test_score  split1_test_score  split2_test_score  mean_test_score  \
0              0.700                0.7               0.70         0.700000   
1              0.700                0.7               0.70         0.700000   
2              0.925                1.0               0.95         0.958333   
3              0.925                1.0               0.95         0.958333   
4              0.975                1.0               0.95         0.975000   
5              0.975                1.0               0.95         0.975000   

   std_test_score  rank_test_score  
0    1.110223e-16                5  
1    1.110223e-16                5  
2    3.118048e-02                3  
3    3.118048e-02                3  
4    2.041241e-02                1  
5    2.041241e-02                1

## 2) best_params_, grid_tree.best_score_
: 최적 파라미터, 최고 정확도

In [8]:
grid_tree.best_params_

{'max_depth': 3, 'min_samples_split': 2}


grid_parameters = {'max_depth':[1,2,3], 'min_samples_split': [2,3]} 에서 

격자로 곱해줘서 가장 좋은걸 찾아줌

In [10]:
grid_tree.best_score_

0.975

In [11]:
pred = grid_tree.predict(X_test)
pred

array([1, 2, 1, 0, 0, 1, 1, 1, 1, 2, 2, 1, 1, 0, 0, 2, 1, 0, 2, 0, 2, 2,
       1, 1, 1, 1, 0, 0, 2, 2])

In [12]:
accuracy_score(y_test, pred)

0.9666666666666667

# < 피처 스케일링 >
- feature scaling
- 서로 다른 변수의 값 범위를 일정한 수준으로 맞추는 작업
- 대표적인 방법으로 1) 표준화 2) 정규화가 있다.

## → 표준화(Standardization)

In [16]:
iris = iris['data']
iris_df = pd.DataFrame(iris)
iris_df

0    1    2    3
0    5.1  3.5  1.4  0.2
1    4.9  3.0  1.4  0.2
2    4.7  3.2  1.3  0.2
3    4.6  3.1  1.5  0.2
4    5.0  3.6  1.4  0.2
..   ...  ...  ...  ...
145  6.7  3.0  5.2  2.3
146  6.3  2.5  5.0  1.9
147  6.5  3.0  5.2  2.0
148  6.2  3.4  5.4  2.3
149  5.9  3.0  5.1  1.8

[150 rows x 4 columns]

In [25]:
# standardize( )으로 함수 만들기
def standardize(x):
    return (x - x.mean()) / x.std()

In [18]:
iris_df.apply(standardize).mean()  

0   -1.457168e-15
1   -1.638319e-15
2   -1.292300e-15
3   -5.543714e-16
dtype: float64

평균이 0이 나오면 잘 나온것 

e-16 이런식으로 나오면 0으로 나온것임.

### ▶ StandardScaler() : 한 번에 표준화

In [19]:
from sklearn.preprocessing import StandardScaler

In [23]:
scaler = StandardScaler()
scaler.fit(iris_df)
iris_scaled = scaler.transform(iris_df)
iris_scaled

array([[-9.00681170e-01,  1.01900435e+00, -1.34022653e+00,
        -1.31544430e+00],
       [-1.14301691e+00, -1.31979479e-01, -1.34022653e+00,
        -1.31544430e+00],
       [-1.38535265e+00,  3.28414053e-01, -1.39706395e+00,
        -1.31544430e+00],
       [-1.50652052e+00,  9.82172869e-02, -1.28338910e+00,
        -1.31544430e+00],
       [-1.02184904e+00,  1.24920112e+00, -1.34022653e+00,
        -1.31544430e+00],
       [-5.37177559e-01,  1.93979142e+00, -1.16971425e+00,
        -1.05217993e+00],
       [-1.50652052e+00,  7.88807586e-01, -1.34022653e+00,
        -1.18381211e+00],
       [-1.02184904e+00,  7.88807586e-01, -1.28338910e+00,
        -1.31544430e+00],
       [-1.74885626e+00, -3.62176246e-01, -1.34022653e+00,
        -1.31544430e+00],
       [-1.14301691e+00,  9.82172869e-02, -1.28338910e+00,
        -1.44707648e+00],
       [-5.37177559e-01,  1.47939788e+00, -1.28338910e+00,
        -1.31544430e+00],
       [-1.26418478e+00,  7.88807586e-01, -1.22655167e+00,
      

### ▶ MinMaxScaler() : 데이터값을 0,1로 변환

In [27]:
from sklearn.preprocessing import MinMaxScaler

In [28]:
scaler = MinMaxScaler()
scaler.fit(iris_df)
iris_scaled = scaler.transform(iris_df)
iris_scaled

array([[0.22222222, 0.625     , 0.06779661, 0.04166667],
       [0.16666667, 0.41666667, 0.06779661, 0.04166667],
       [0.11111111, 0.5       , 0.05084746, 0.04166667],
       [0.08333333, 0.45833333, 0.08474576, 0.04166667],
       [0.19444444, 0.66666667, 0.06779661, 0.04166667],
       [0.30555556, 0.79166667, 0.11864407, 0.125     ],
       [0.08333333, 0.58333333, 0.06779661, 0.08333333],
       [0.19444444, 0.58333333, 0.08474576, 0.04166667],
       [0.02777778, 0.375     , 0.06779661, 0.04166667],
       [0.16666667, 0.45833333, 0.08474576, 0.        ],
       [0.30555556, 0.70833333, 0.08474576, 0.04166667],
       [0.13888889, 0.58333333, 0.10169492, 0.04166667],
       [0.13888889, 0.41666667, 0.06779661, 0.        ],
       [0.        , 0.41666667, 0.01694915, 0.        ],
       [0.41666667, 0.83333333, 0.03389831, 0.04166667],
       [0.38888889, 1.        , 0.08474576, 0.125     ],
       [0.30555556, 0.79166667, 0.05084746, 0.125     ],
       [0.22222222, 0.625     ,

In [30]:
 #이렇게 써줘도 결과는 똑같다.
iris_scaled = scaler.fit_transform(iris_df)
iris_scaled

array([[0.22222222, 0.625     , 0.06779661, 0.04166667],
       [0.16666667, 0.41666667, 0.06779661, 0.04166667],
       [0.11111111, 0.5       , 0.05084746, 0.04166667],
       [0.08333333, 0.45833333, 0.08474576, 0.04166667],
       [0.19444444, 0.66666667, 0.06779661, 0.04166667],
       [0.30555556, 0.79166667, 0.11864407, 0.125     ],
       [0.08333333, 0.58333333, 0.06779661, 0.08333333],
       [0.19444444, 0.58333333, 0.08474576, 0.04166667],
       [0.02777778, 0.375     , 0.06779661, 0.04166667],
       [0.16666667, 0.45833333, 0.08474576, 0.        ],
       [0.30555556, 0.70833333, 0.08474576, 0.04166667],
       [0.13888889, 0.58333333, 0.10169492, 0.04166667],
       [0.13888889, 0.41666667, 0.06779661, 0.        ],
       [0.        , 0.41666667, 0.01694915, 0.        ],
       [0.41666667, 0.83333333, 0.03389831, 0.04166667],
       [0.38888889, 1.        , 0.08474576, 0.125     ],
       [0.30555556, 0.79166667, 0.05084746, 0.125     ],
       [0.22222222, 0.625     ,

#### < 주의 : fit >

X_train, X_test = train_test_split(iris_data)

X_train

scaler.fit()
scaler.fit_transform(X_train)

1,2,1,2 ---> 10

X_test

scaler.transform(x_test)
1,2,1,2 ----> 10? X

먼저 있는것을 정규화 시키고 쪼개라 

쪼개놓고 각각 정규화 시키면 위의 fit기준과 아래의 fit기준이 달라지게 된다.

즉, fit은 한번만 해줘야한다.

< 올바른 표현 >

scaler.fit(x)
scaler.transformm(x)

## → 정규화(Normalization)

In [32]:
from sklearn.datasets import load_iris
iris = load_iris()

iris = iris['data']
iris_df = pd.DataFrame(iris)
iris_df

0    1    2    3
0    5.1  3.5  1.4  0.2
1    4.9  3.0  1.4  0.2
2    4.7  3.2  1.3  0.2
3    4.6  3.1  1.5  0.2
4    5.0  3.6  1.4  0.2
..   ...  ...  ...  ...
145  6.7  3.0  5.2  2.3
146  6.3  2.5  5.0  1.9
147  6.5  3.0  5.2  2.0
148  6.2  3.4  5.4  2.3
149  5.9  3.0  5.1  1.8

[150 rows x 4 columns]

In [33]:
iris_df.max()

0    7.9
1    4.4
2    6.9
3    2.5
dtype: float64

In [34]:
iris_df.min()

0    4.3
1    2.0
2    1.0
3    0.1
dtype: float64

In [35]:
def min_max(x):
    return (x - x.min()) / (x.max() - x.min())

In [36]:
iris_df.apply(min_max)

0         1         2         3
0    0.222222  0.625000  0.067797  0.041667
1    0.166667  0.416667  0.067797  0.041667
2    0.111111  0.500000  0.050847  0.041667
3    0.083333  0.458333  0.084746  0.041667
4    0.194444  0.666667  0.067797  0.041667
..        ...       ...       ...       ...
145  0.666667  0.416667  0.711864  0.916667
146  0.555556  0.208333  0.677966  0.750000
147  0.611111  0.416667  0.711864  0.791667
148  0.527778  0.583333  0.745763  0.916667
149  0.444444  0.416667  0.694915  0.708333

[150 rows x 4 columns]

In [37]:
iris_df.apply(min_max).max()

0    1.0
1    1.0
2    1.0
3    1.0
dtype: float64

In [38]:
#람다식을 이용한 같은 표현
iris_df.apply(lambda x:(x - x.min()) / (x.max() - x.min()))

0         1         2         3
0    0.222222  0.625000  0.067797  0.041667
1    0.166667  0.416667  0.067797  0.041667
2    0.111111  0.500000  0.050847  0.041667
3    0.083333  0.458333  0.084746  0.041667
4    0.194444  0.666667  0.067797  0.041667
..        ...       ...       ...       ...
145  0.666667  0.416667  0.711864  0.916667
146  0.555556  0.208333  0.677966  0.750000
147  0.611111  0.416667  0.711864  0.791667
148  0.527778  0.583333  0.745763  0.916667
149  0.444444  0.416667  0.694915  0.708333

[150 rows x 4 columns]